main.cpp

```diff
@@ -28,22 +28,22 @@ struct T
 
 struct Struct1                                //4
 {
-    const T* compare(const T& a,const T& b) //5
+    // This is a valid code but your instruction is "convert all pointer to reference type where possible"
+    const T* compare(const T& a, const T& b) //5
     {
-        if( a.value < b.value ) return &a;
+        if( a.value < b.value ) return &b;
         if( a.value > b.value ) return &a;
-        return &a;   // Added to remove the compiler warnings "control may reach end of non-void function [-Wreturn-type]"
+        return nullptr;   
     }
 
-    T* compare2(T& a, T& b) //5
+    // This is hopefully the right answer for your instruction 
+    const T& compare3(const T& a, const T& b) //5
     {
-        if( a.value < b.value ) return &a;
-        if( a.value > b.value ) return &a;
-        return &a;   // Added to remove the compiler warnings "control may reach end of non-void function [-Wreturn-type]"
+        if( a.value < b.value ) return b;
+        if( a.value > b.value ) return a;
+        return a;   // If a and b is equal. Not the best solution...
```

Comment:
This is the question I asked:
`can the return type T* be converted to a reference? hint: you're returning a nullptr at the end.`
Here is your original implementation:
```
    T* compare(T* a, T* b) //5
    {
        if(a != nullptr && b != nullptr) 
        {
            if( a->value < b->value ) return a;
            if( a->value > b->value ) return b;
        }
        return nullptr;
    }
```
If we study this function, we can see that it returns a `nullptr` if: 
- the 2 inputs have the same value. 
- one or both of the inputs are `nullptr`.

You wrote the following in response to the above question: 
`yes, it's possible because in the scope of the member function exists valid references of a and b`

First, `a` and `b` are not member functions anywhere.  They are arguments to the function. 
It just so happens that in `main()` `a` and `b` have the following values: 
```
    T t1(10 , "a");    //this is 'a' inside 'compare()'.  its value is '10'
    T t2(20 , "b");    //this is 'b' inside 'compare()'.  its value is '20'
    
    Struct1 f;                                            //7
    auto* smaller = f.compare(&t1 , &t2); //here they are passed and become 'a' and 'b' inside 'compare()'
```
`t1` becomes the `a` inside the `compare` function, and `t2` becomes the `b` inside the function.
`t1`'s value (or `a->value`) is 10.  `t2`'s value (or `b->value`) is 20. 
We can force the `compare` function to return a `nullptr` by making `t1` and `t2` have the same value or by passing `nullptr` to either function argument. 

Once this `compare()` function is changed to use references as the arguments, it is no longer possible to pass in `nullptr` for either argument.
However it is **still possible** for both arguments to have the **same value**.
Consider this revision to `main()`:
```
    T t1(10 , "a");  //this is 'a' inside 'compare()'.  it's value is '10'
    T t2(10 , "b");  //this is 'b' inside 'compare()'   it's value is '10'
    
    Struct1 f;                                            //7
    auto* smaller = f.compare(&t1 , &t2); 
```
Now that `t1` and `t2` have the same value, what should the `compare()` function return? 
The implementation from `Part1` returned a `nullptr` if `a` and `b` have the same value. 

Your `compare()` function **still needs to return a `nullptr`** if the inputs have the same value.  
**You need to preserve that functionality** when you convert the pointers to references. 
The instructions give you this hint: 
```
        hint: There is no reference equivalent to nullptr.  
        if a pointer (including nullptr) is being returned anywhere, 
             don't try to convert it to a reference.
```
Thus, **your answer** to this question `can the return type T* be converted to a reference?` **is wrong**:

`yes, it's possible because in the scope of the member function exists valid references of a and b`

The correct answer is:
```
NO.  The return type T* can NOT be converted to a reference. 
```

The `compare()` function needs to return a `nullptr` if the two inputs have the same value.
Your revised implementation forces the function to no longer have that ability. 
This is the correct revision of the `compare` function, with lots of comments added to clarify how each line was changed:
```
    // the argument type T* can be safely converted to a reference: T&. 
    // They do not need to be 'const' either as the original return type was 'T*', not 'const T*'

    T* compare(T& a, T& b)  
    {
        //now that the arguments are references, we don't have to check for nullptr.
        //if(a != nullptr && b != nullptr)  
        //{

        //the function expects a pointer to be returned. return the address of 'a'
        if( a.value < b.value ) return &a; 
        
        //the function expects a pointer to be returned. return the address of 'b'
        if( a.value > b.value ) return &b; 

        //}
        
        //if a pointer (including nullptr) is being returned anywhere, don't try to convert it to a reference.
        return nullptr; 
    }
```
Please make this revision (omit the commented lines) and resubmit your code for review.
- [ ] revise